core/http-auth/src/main/java/software/amazon/awssdk/http/auth/internal/util/CanonicalRequestV2.java

```diff
@@ -0,0 +1,329 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.auth.internal.util;
+
+import static software.amazon.awssdk.utils.StringUtils.lowerCase;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.SortedMap;
+import java.util.TreeMap;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.http.SdkHttpRequest;
+import software.amazon.awssdk.utils.Pair;
+import software.amazon.awssdk.utils.StringUtils;
+import software.amazon.awssdk.utils.http.SdkHttpUtils;
+
+/**
+ * A class that represents a canonical request in AWS, as documented
+ * here: <a href="https://docs.aws.amazon.com/IAM/latest/UserGuide/create-signed-request.html#create-canonical-request">...</a>
+ */
+@SdkInternalApi
+public final class CanonicalRequestV2 {
+    private static final List<String> HEADERS_TO_IGNORE_IN_LOWER_CASE =
+        Arrays.asList("connection", "x-amzn-trace-id", "user-agent", "expect");
+
+    private final SdkHttpRequest request;
+    private final String contentHash;
+    private final Options options;
+
+    private final String canonicalUri;
+    private final SortedMap<String, List<String>> canonicalParams;
+    private final List<Pair<String, List<String>>> canonicalHeaders;
+    private final String canonicalParamsString;
+    private final String canonicalHeadersString;
+    private final String signedHeadersString;
+    private final String canonicalRequestString;
+
+    public CanonicalRequestV2(SdkHttpRequest request, String contentHash, Options options) {
+        this.request = request;
+        this.contentHash = contentHash;
+        this.options = options;
+
+        this.canonicalParams = getCanonicalQueryParams(request);
+        this.canonicalHeaders = getCanonicalHeaders(request.headers());
+        this.canonicalUri = getCanonicalUri(request, options);
+        this.canonicalParamsString = getCanonicalQueryString(canonicalParams);
+        this.canonicalHeadersString = getCanonicalHeadersString(canonicalHeaders);
+        this.signedHeadersString = getSignedHeadersString(canonicalHeaders);
+        this.canonicalRequestString = getString();
+    }
+
+    /**
+     * Get the canonical request string for a given {@link SdkHttpRequest}, content-hash, and {@link Options}.
+     */
+    public static String getString(SdkHttpRequest request, String contentHash, Options options) {
```

Comment:
I think remove this

Reply:
Yep. It should use parameters of a canonical request instead of the params of the constructor. Updated.